usaspending_api/common/threaded_data_loader.py

```diff
@@ -0,0 +1,194 @@
+        # Make sure we weren't flagged to exit
+        while not self.loader.exit_flag:
+            if not self.data_queue.empty():
+                row = self.data_queue.get()
+                # Grab the collision field
+                update = False
+                collision_instance = None
+                model_instance = None
+                if self.loader.collision_field is not None:
+                    model_collision_field = self.loader.collision_field
+                    data_collision_field = model_collision_field
+                    # If it's a long label, we need to look it up to map to data
+                    if model_collision_field in settings.LONG_TO_TERSE_LABELS:
+                        data_collision_field = settings.LONG_TO_TERSE_LABELS[model_collision_field]
+                    query_object = Q({model_collision_field: row[data_collision_field]})
+                    collision_instance = self.loader.model_class.objects.filter(query_object).first()
+
+                    if collision_instance is not None:
+                        behavior = self.loader.collision_behavior
+                        if behavior is "delete":  # Delete the row from the data store and load new row
+                            collision_instnace.delete()
+                        if behavior is "update":  # Update the row in the data store with new values from the CSV
+                            update = True
+                        if behavior is "skip":  # Skip the row
+                            continue
+                        if behavior is "skip_and_complain":  # Log a warning and skip the row
+                            print('Hit a collision on row %s' % (row))
+                        if behavior is "die":  # Raise an exception and cease execution
+                            raise Exception("Hit collision on row %s" % (row))
+
+                if not update:
+                    model_instance = self.loader.model_class()
+                else:
+                    model_instance = collision_instance
+
+                if self.loader.pre_row_function is not None:
+                    self.loader.pre_row_function(row=row, instance=model_instance)
+
+                try:
+                    processed_row = self.load_data_into_model(model_instance,
+                                                              self.loader.fields,
+                                                              self.loader.field_map,
+                                                              self.loader.value_map,
+                                                              row)
+                except Exception as e:
+                    print(e)
+                # If we have a post row function, run it before saving
+                if self.loader.post_row_function is not None:
+                    self.loader.post_row_function(row=row, instance=model_instance)
+
+                model_instance.save()
+            time.sleep(1)
```

Comment:
To prevent thread lock when accessing self.load.exit_flag (without it you might run into a situation where a single thread is constantly checking that variable and holding priority)